### DISCUSSION TITLE
HydroDyn/Current CurrMod2 how to complie

### DISCUSSION THREAD

**Discussion by LeoHongLiZuo** [OpenFAST/openfast#1517] [Q&A]
Dear guys
I want to use Current(CurrMod=2 user-defined from routine UserCurrent）
I have find the modified location after asking.
As shown in the following figure.
![图片](https://user-images.githubusercontent.com/108517989/230526185-42353add-f02f-4056-b168-0801d3a43454.png)
But I'm not familiar with how to complie... 
May I ask who  have any examples for reference or how to compile it. I want to learn about how to compile it.

**Comment by jjonkman** [comment]
Dear @LeoHongLiZuo,

Documentation on compiling OpenFAST is available here: https://openfast.readthedocs.io/en/main/source/install/index.html#compile-from-source.  The easiest approach depends on what operating system you have.

Best regards,

**Reply by LeoHongLiZuo** [reply]
Dear Jonkman,
I‘m very grateful to you for your advice! I've been trying to solve this problem for the past few days.
But I still want to ask you about some of it.
Firstly,  I will follow Current.f90 guidance, put the  USER-DEFINED ROUTINE (UserCurrent) in the UserSubs.f90.
And the parameters are set as shown below.
![图片](https://user-images.githubusercontent.com/108517989/230902397-ffdf18db-6c9a-4673-a35b-05cf0ed77250.png)
But after I saving and setting CurrMod(2user-defined), running the Openfast the results doesn't change.
I don't know if it is because I need to recompile the FAST.sln after setting the parameters, or I set the parameters in the wrong format.
If the format is wrong, I found the Current.txt
![图片](https://user-images.githubusercontent.com/108517989/230903421-b51b0959-1372-4dc7-9938-0a15e822fa09.png)
But I don't think I can refer to this file to write the standard format...
Secondly,I would like to ask the definition of a parameter.
![图片](https://user-images.githubusercontent.com/108517989/230903984-d4d7be23-10e7-413d-97c2-399580c10092.png)
For the CurrNSRef in the picture.I wonder if it mean the depth from sea level to sea floor? If so, why does it start with 20 instead of 130? Or this parameter refers to the reference depth from set the CURRENT plane to sea level?
Please forgive my lack of ability and thank you very much for your help!
But I really want to learn Openfast well and I'm willing to put in the effort.
Best regards!

**Reply by jjonkman** [reply]
Dear @LeoHongLiZuo,

After changing `SUBROUTINE UserCurrent()` in _Current.f90_, you must recompile OpenFAST to make use of the changes.

But your implementation of `UserCurrent()` is not correct.  `WtrDpth` and `zi` are `INTENT(IN)`, meaning they are passed into the `SUBROUTINE` for you to use, but you cannot change their values.  You must use these values to calculate `CurrVxi` and `CurrVyi` at the reference depth of `zi`.

The near-surface current in HydroDyn is a simple linear current shear with a velocity of `CurrNSV0` at the still water level down to zero at a depth of `CurrNSRef` (and is meant for near-surface currents where `CurrNSRef` < `WtrDpth`).  For more information on the built-in current options available in HydroDyn, see the online HydroDyn documentation here: https://openfast.readthedocs.io/en/main/source/user/hydrodyn/input_files.html#current.

Best regards,

**Reply by LeoHongLiZuo** [reply]
Dear Jonkman,
Thank you very much for hearing from you again！
I tried compiling Openfast again over the last few days and it worked.
There is one question I would like to ask you.
After I repeated many times to make sure there were no errors in the steps.
As shown in the figure below, MAP_x64.dll was successfully compiled when openfast-2.4.0 was compiled.
![图片](https://user-images.githubusercontent.com/108517989/231770041-6e36f2c3-0c2e-40a1-ab50-2ae9fead810d.png)
However, the figure below shows that no MAP_x64.dll was generated during the compilation of  Openfast3.4.1.
![图片](https://user-images.githubusercontent.com/108517989/231770460-49958dd9-f0be-4642-ac26-8506a9fbe130.png)
But it looks like MAP_x64.dll is the necessary file to run openfast_x64.exe.
I'm not sure if it's because I made a mistake compiling Openfast3.4.1, or if it already has MAP_x64.dll in it for the Openfast3.4.1 version of openfast_x64.exe.
Best regards!

**Reply by jjonkman** [reply]
Dear @LeoHongLiZuo,

As of OpenFAST v3.1 and newer, MAP++ is now statically linked within the OpenFAST executable, so, a standalone MAP++ dynamic library is no longer needed.

Best regards,

**Reply by LeoHongLiZuo** [reply]
Dear Jonkman,
Well, I see. Thank you very much for your patient guidance and help!
Best regards!
